src/Catel.Core/Catel.Core.Shared/IoC/TypeFactory.cs

```diff
@@ -296,7 +296,7 @@ private void InitializeAfterConstruction(object obj, TypeMetaData typeMetaData)
 
                 Log.Debug("Creating instance of type '{0}' using specific parameters. No constructor found in the cache, so searching for the right one", typeToConstruct.FullName);
 
-                var constructors = typeConstructorsMetadata.GetConstructors(parameters.Length, !autoCompleteDependencies);
+                var constructors = typeConstructorsMetadata.GetConstructors(parameters.Length, !autoCompleteDependencies).SortByParametersMatchDistance(parameters).ToList();
```

Comment:
var constructors = typeConstructorsMetadata.GetConstructors(parameters.Length, !autoCompleteDependencies);
if (constructors.Count > 1)
{
    // only do the expensive sorting here
}